lib/ex_doc/formatter/html/templates.ex

```diff
@@ -30,6 +30,27 @@ defmodule ExDoc.Formatter.HTML.Templates do
     nil
   end
 
+  @doc """
+  Get the attribute string used to define the spec of the given `node`.
+  """
+  def get_spec_attribute(module, node) do
+    node
+    |> get_spec_attribute()
+    |> module.language.format_attribute()
+  end
+
+  def get_spec_attribute(%ExDoc.TypeNode{} = node) do
+    if "opaque" in node.annotations do
+      "opaque"
+    else
+      "type"
+    end
+  end
+
+  def get_spec_attribute(%ExDoc.FunctionNode{type: :callback}), do: "callback"
+  def get_spec_attribute(%ExDoc.FunctionNode{type: :macrocallback}), do: "macrocallback"
```

Comment:
macrocallback is Elixir specific so ideally it wouldn't be in a language-agnostic module like the html formatter.

The new `format_attribute` callback make sense within the existing code structure where we store the spec ast _without_ the _kind_ (kind being basically `@type`, `-type`, etc). However, if we change the internals to do store the kind in ast, we wouldn't need the callback. In fact if we do so, we wouldn't need the logic that trims the kind in Erlang: https://github.com/elixir-lang/ex_doc/blob/main/lib/ex_doc/language/erlang.ex#L588:L589.

Reply:
The best I can do so far is to move these functions into the language modules so that `get_spec_attribute(module, node)` becomes:

```elixir
def get_spec_attribute(module, node) do
  module.language.get_spec_attribute(node)
end
```

And those respective functions return either "@type" or "-type" as appropriate. With `:macrocallback` not present in the Erlang module of course. Let me know if I should update the PR with this change.

I looked into including the original text in the AST but kept hitting issues, so would need some hints for that. I'm not sure how you would want to keep the kind from Erlang while still having it separated so that the output can be wrapped in the formatting span element?

Reply:
> I'm not sure how you would want to keep the kind from Erlang while still having it separated so that the output can be wrapped in the formatting span element?

yeah, good point. I didn't realise we need the kind separate so we can render it with appropriate css class.

Your snippet looks good, I'd just rename it to `format_spec_attribute` (both function & callback)